### PR TITLE
Create Base API

### PR DESCRIPTION
**Pull request by xuanthu2002**
## Related Tickets
- [#Task 78533: Create Base API](https://edu-redmine.sun-asterisk.vn/issues/78533)
## WHAT
- Create Base API
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/blob/master/eng/android/coding_convention_android_kotlin.md | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [ ] yes</li>|<li>- [ ] yes</li>
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(Other notes)*

### CODE REVIEW

**Review by MaxHolmes221B — APPROVED** [review]

**Review by TranTinh030892 — APPROVED** [review]

File: app/src/main/java/com/example/fina/data/model/Coin.kt [hunk 1]
```diff
@@ -1,4 +1,91 @@
...
+    }
+}
+
+enum class OrderDirection(private val displayName: String) {
+    DESC("desc"),
+    ASC("asc");
+
+    override fun toString(): String {
+        return displayName
+    }
```
Comment (ghost) [inline app/src/main/java/com/example/fina/data/model/Coin.kt]: Tách các model thành các file riêng em ạ, không ôm đồm thế này, enum thì nên move sang 1 chỗ riêng (constant)

File: app/src/main/java/com/example/fina/data/repository/source/CoinDataSource.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import com.example.fina.data.model.OrderBy
+import com.example.fina.data.model.OrderDirection
+import com.example.fina.data.model.PriceHistory
+import com.example.fina.data.model.TimePeriod
+import com.example.fina.data.repository.OnResultListener
+import com.example.fina.utils.Constant
+
+interface CoinDataSource {
+    interface Local {
+        fun getLocalCoins(listener: OnResultListener<MutableList<Coin>>)
```
Comment (ghost) [inline app/src/main/java/com/example/fina/data/repository/source/CoinDataSource.kt]: Dùng List thay vì MutableList em ơi, Data từ tầng Source trả về là cố định mà
```suggestion
        fun getLocalCoins(listener: OnResultListener<List<Coin>>)
```

File: app/src/main/java/com/example/fina/data/repository/source/CoinDataSource.kt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    interface Remote {
+        fun getCoins(
+            referenceCurrencyUuid: String = Constant.USD_UUID,
+            timePeriod: TimePeriod = TimePeriod.TWENTY_FOUR_HOURS,
+            uuids: MutableList<String> = mutableListOf(),
+            search: String? = null,
+            orderBy: OrderBy = OrderBy.MARKET_CAP,
+            orderDirection: OrderDirection = OrderDirection.DESC,
+            limit: Int = Constant.DEFAULT_LIMIT,
+            offset: Int = 0,
```
Comment (ghost) [inline app/src/main/java/com/example/fina/data/repository/source/CoinDataSource.kt]: Nhiều tham số quá, sao em pass được detekt vậy???
Em có thể gom các param liên quan vào các class. Ví dụ: OrderProperties chứa cả OrderBy và OrderDirection, limit, offset, ....

File: app/src/main/java/com/example/fina/data/repository/source/remote/CoinRemoteDataSource.kt [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    override fun getCoins(
+        referenceCurrencyUuid: String,
+        timePeriod: TimePeriod,
+        uuids: MutableList<String>,
+        search: String?,
+        orderBy: OrderBy,
+        orderDirection: OrderDirection,
+        limit: Int,
+        offset: Int,
```
Comment (ghost) [inline app/src/main/java/com/example/fina/data/repository/source/remote/CoinRemoteDataSource.kt]: Nhiều tham số quá e

**Review by ghost — CHANGES_REQUESTED** [review]
check my cmts

**Review by ghost — APPROVED** [review]

**Pull request merged**